packages/react-router/src/router.ts

```diff
@@ -956,65 +956,94 @@ export class Router<
   ```ts
   matchRoutes (
     next: ParsedLocation,
-    opts?: { preload?: boolean; throwOnError?: boolean },
+    opts?: MatchRoutesOpts,
+    routeFullPath?: string,
   ): Array<AnyRouteMatch>;
   ```
 */
   public matchRoutes(
     pathname: string,
     locationSearch: AnySearchSchema,
     opts?: MatchRoutesOpts,
+    routeFullPath?: string,
   ): Array<AnyRouteMatch>
   public matchRoutes(
     next: ParsedLocation,
     opts?: MatchRoutesOpts,
+    routeFullPath?: string,
   ): Array<AnyRouteMatch>
 
   public matchRoutes(
     pathnameOrNext: string | ParsedLocation,
-    locationSearchOrOpts?:
-      | AnySearchSchema
-      | { preload?: boolean; throwOnError?: boolean },
-    opts?: { preload?: boolean; throwOnError?: boolean },
+    locationSearchOrOpts?: AnySearchSchema | MatchRoutesOpts,
+    optsOrRouteFullPath?: MatchRoutesOpts | string,
+    routeFullPath?: string,
   ) {
     if (typeof pathnameOrNext === 'string') {
       return this.matchRoutesInternal(
         {
           pathname: pathnameOrNext,
           search: locationSearchOrOpts,
         } as ParsedLocation,
-        opts,
+        optsOrRouteFullPath as MatchRoutesOpts,
+        routeFullPath,
       )
     } else {
-      return this.matchRoutesInternal(pathnameOrNext, locationSearchOrOpts)
+      return this.matchRoutesInternal(
+        pathnameOrNext,
+        locationSearchOrOpts,
+        optsOrRouteFullPath as string,
+      )
     }
   }
 
   private matchRoutesInternal(
     next: ParsedLocation,
     opts?: { preload?: boolean; throwOnError?: boolean },
+    routeFullPath?: string,
   ): Array<AnyRouteMatch> {
     let routeParams: Record<string, string> = {}
 
-    const foundRoute = this.flatRoutes.find((route) => {
+    let foundRoute: AnyRoute | undefined = routeFullPath
+      ? this.routesByPath[routeFullPath]
+      : undefined
```

Comment:
does this mean if `to` is not specified, the link will still be slow?

in this case, can we use the current location of router instead?

Reply:
Is it a valid use case to use a Link without `to`? If yes, we could provide the current match's `fullPath` as a fallback. 

Currently, it only speeds up the resolve process for links with absolute paths, because they can be easily looked up in the `routeById` map. Is there any util to quickly resolve relative paths?

Reply:
> Is it a valid use case to use a Link without to?
yes, see e.g. https://tanstack.com/router/latest/docs/framework/react/guide/search-params#link-search-

in short: if you omit `to` but specify `from`, you will get typesafe links that point to the current location and can be used for e.g. updating search params.

Reply:
> Is there any util to quickly resolve relative paths?

we have https://github.com/TanStack/router/blob/main/packages/react-router/src/path.ts#L92
whether that's fast should be benchmarked